### PR TITLE
POC: Spike GraphQL query management and merging strategy (DONT MERGE)

### LINKED ISSUE #30790: Spike: GraphQL Query Management and Merging Strategy
# Purpose
Depending on the results of https://github.com/dotCMS/core/issues/30695 this one can change.
Research and document GraphQL integration challenges for the dotCMS SDK.

### 1. GraphQL Page API Analysis

- Permission model
- Error handling patterns
- Data structure limitations
- Query depth restrictions (if any, I think none)
- Field validation
- Cache implications
- More…?

### 2. Query Merging Solutions

- Fragment composition
- Schema stitching
- Query batching
- Field conflicts resolution
- Type extension handling

### GraphQL API Investigation
Create test cases for basic queries, permission-based queries, and error scenarios.

```javascript
// Ideas of test cases, this is oversimplified, we need more
const testQueries = [
    // Basic page query
    `query {
        page(url: "/test") {
            title
            layout
        }
    }`,
    
    // With permissions
    `query {
        page(url: "/test", mode: EDIT) {
            title
            layout
        }
    }`,
    
    // Error cases
    `query {
        page(url: "/invalid") {
            invalid_field
        }
    }`
];
```
## Query Merging POC

Design a strategy for merging default and custom GraphQL queries.
Validate merged queries and resolve potential field conflicts.

```javascript
interface QueryMerger {
    merge(defaultQuery: string, customQuery: string): string;
    validate(mergedQuery: string): boolean;
    resolveConflicts(fields: string[]): string[];
}

// Test scenarios
const scenarios = [
    {
        default: `
            query Page($url: String!) {
                page(url: $url) {
                    layout
                    template
                }
            }
        `,
        custom: `
            query PageExtra {
                seo {
                    title
                }
            }
        `
    }
];
```

### Proposed Objective

Technical User Experience

### Proposed Priority

Priority 3 - Average

## GraphQL API Research

- Document permission model
- Document the GQL features not supported or limited (if any)
- List all possible error types
- Map field restrictions
- Document query depth limits
- Test caching behavior

## Query Merging

- Test different merging strategies
- Document conflict resolution
- Create working POC
- Performance impact analysis

## Documentation

- API limitations guide
- Query merging best practices
- Error handling patterns
- Performance recommendations

### External Links... Slack Conversations, Support Tickets, Figma Designs, etc.

We have a document where we are framing the main objective and the reasons of this Spike. Feel free to reach me if you want access. 

### Assumptions & Initiation Needs

- You understand how GQL queries work 
- You have knowledge about dotCMS GQL Schemas 
- You have knowledge of which fields per type of content are required in the SDK in order to work

### PR DESCRIPTION
**Pull request by zJaaal**
# Summary 

This POC makes use of [GraphQL Fragments](https://graphql.org/learn/queries/#fragments) in order to merge queries. Also makes use of the proposed API that we selected in the planning stage.

```javascript
const data = await client.gql({
        page: {
            url: params?.slug?.join('/'),
            language: searchParams?.language_id,
            mode: 'EDIT_MODE',
            pageFragment: `
            containers {
                containerContentlets {
                    contentlets {
                        ... on Blog {
                            author {
                                firstName
                                lastName
                            }
                        }
                    }
                }
            }
        `
        },
        content: {
            blogs: `search(query: "+contentType: blog", limit: 3) {
            _map
            title
            ...on Blog {
                author {
                    title
                }
            }
        }`,
            destinations: `search(query: "+contentType: destination", limit: 3) {
            _map
        }`
        }
    });
```

This way the developer should only write a `Fragment` to extend the Page API response and also can write as much queries as they want inside the `content` object.

In this case all limitations, errors, query depth, cache handling and implications will be inherited naturally from how the GraphQL API works, since we are using a native way of merge queries and should not require any extra work from our side apart from exposing a simple way of using dotCMS API with GQL Queries and adapt our current architecture to make it editable inside UVE (which changes can be seen in this PR in a simple version of how it should be).

### Content Editing with this POC on UVE

https://github.com/user-attachments/assets/1b82661c-c194-4e14-a929-2c057f31b7dd

This POC is just a proof of how this can be achieved and not a final design of how this will work or the architecture of it, we still have to make some work to make this as clean, easy and reuse as much code as we can (building the queries inside content, at least the Lucene Query, can be done with an existing API).

### CODE REVIEW

File: core-web/libs/sdk/client/src/lib/client/gql/utils.ts [hunk 1]
```diff
@@ -0,0 +1,82 @@
+export const buildQuery = ({
+    pageFragment,
+    contentQueries
+}: {
+    pageFragment: string;
+    contentQueries: string;
+}) => `
+  fragment DotCMSPage on DotPage {
+    _map
```
Comment (fmontes) [inline core-web/libs/sdk/client/src/lib/client/gql/utils.ts:9]: Don't think we need this.

File: core-web/libs/sdk/client/src/lib/client/gql/utils.ts [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+          }
+        }
+      }
+    }
+    viewAs {
+      visitor {
+        persona {
+          name
+        }
+      }
```
Comment (fmontes) [inline core-web/libs/sdk/client/src/lib/client/gql/utils.ts:53]: Do we need this?

File: examples/nextjs/src/app/[[...slug]]/page.js [hunk 3]
```diff
@@ -26,33 +26,71 @@ export async function generateMetadata({ params, searchParams }) {
...
         };
     };
-    const { pageAsset, nav, error } = await getPageData();
+    const { nav, error } = await getNavigation();
+
+    const data = await client.gql({
+        page: {
+            url: params?.slug?.join('/'),
+            language: searchParams?.language_id,
+            mode: 'EDIT_MODE',
```
Comment (fmontes) [inline examples/nextjs/src/app/[[...slug]]/page.js:59]: We need to control this, should be optional for them.

File: examples/nextjs/src/components/my-page.js [hunk 4]
```diff
@@ -1,24 +1,24 @@
...
+import Footer from './layout/footer/footer';
+import Navigation from './layout/navigation';
+import { usePathname, useRouter } from 'next/navigation';
+import { DotcmsLayout } from '@dotcms/react';
+import { withExperiments } from '@dotcms/experiments';
+import { CustomNoComponent } from './content-types/empty';
 
-import { usePageAsset } from "../hooks/usePageAsset";
-import NotFound from "@/app/not-found";
+import { usePageAsset } from '../hooks/usePageAsset';
```
Comment (fmontes) [inline examples/nextjs/src/components/my-page.js:20]: Why this keep happening?

File: core-web/libs/sdk/client/src/lib/client/gql/utils.ts (hunk 1 above)
Comment (zJaaal) [reply core-web/libs/sdk/client/src/lib/client/gql/utils.ts:9]: Nope, we can define the props we need.

File: core-web/libs/sdk/client/src/lib/client/gql/utils.ts (hunk 2 above)
Comment (zJaaal) [reply core-web/libs/sdk/client/src/lib/client/gql/utils.ts:53]: Probably not

File: examples/nextjs/src/app/[[...slug]]/page.js (hunk 3 above)
Comment (zJaaal) [reply examples/nextjs/src/app/[[...slug]]/page.js:59]: agree

File: examples/nextjs/src/components/my-page.js (hunk 4 above)
Comment (zJaaal) [reply examples/nextjs/src/components/my-page.js:20]: My global prettier is configured for single quotes and maybe someone else in the team uses double quotes in their IDE, the examples does not have prettier

**Pull request closed without merging**